File: session_cpp/src/arrow_test.cpp
#include "catch_amalgamated.hpp"
#include "arrow.h"

using namespace session_cpp;

TEST_CASE("Arrow creation", "[arrow]") {
    Line line(0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 10.0f);
    Arrow arrow(line, 1.0f);

    REQUIRE(arrow.radius == 1.0f);
    REQUIRE(arrow.mesh.number_of_vertices() == 29);
    REQUIRE(arrow.mesh.number_of_faces() == 28);
    REQUIRE(arrow.name == "my_arrow");
    REQUIRE(!arrow.guid.empty());
}

TEST_CASE("Arrow JSON serialization", "[arrow]") {
    Line line(0.0f, 0.0f, 0.0f, 5.0f, 0.0f, 0.0f);
    Arrow arrow(line, 2.0f);

    auto data = arrow.to_json_data();
    REQUIRE(data["type"] == "Arrow");
    REQUIRE(data["radius"] == 2.0f);
    REQUIRE(data.contains("mesh"));
    REQUIRE(data.contains("line"));
}

TEST_CASE("Arrow JSON round trip", "[arrow]") {
    Line line(1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f);
    Arrow arrow(line, 0.5f);

    std::string filepath = "test_arrow.json";
    arrow.to_json(filepath);

    Arrow loaded = Arrow::from_json(filepath);
    REQUIRE(loaded.radius == 0.5f);
    REQUIRE(loaded.mesh.number_of_vertices() == 29);
    REQUIRE(loaded.mesh.number_of_faces() == 28);
}

TEST_CASE("Arrow mesh has color collections", "[arrow]") {
    Line line(0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 10.0f);
    Arrow arrow(line, 1.0f);

    REQUIRE(arrow.mesh.pointcolors.size() == 29);
    REQUIRE(arrow.mesh.facecolors.size() == 28);
    REQUIRE(arrow.mesh.linecolors.size() == 56);
    REQUIRE(arrow.mesh.widths.size() == 56);
}

File: session_py/src/session_py/arrow_test.py
import pytest
from session_py import Arrow, Line


def test_arrow_creation():
    """Test basic arrow creation."""
    line = Line(0.0, 0.0, 0.0, 0.0, 0.0, 10.0)
    arrow = Arrow(line, 1.0)

    assert arrow.radius == 1.0
    assert arrow.mesh.number_of_vertices() == 29
    assert arrow.mesh.number_of_faces() == 28
    assert arrow.name == "my_arrow"
    assert arrow.guid is not None


def test_arrow_json_serialization():
    """Test arrow JSON serialization."""
    line = Line(0.0, 0.0, 0.0, 5.0, 0.0, 0.0)
    arrow = Arrow(line, 2.0)

    data = arrow.to_json_data()
    assert data["type"] == "Arrow"
    assert data["radius"] == 2.0
    assert "mesh" in data
    assert "line" in data


def test_arrow_json_round_trip():
    """Test arrow JSON file I/O."""
    line = Line(1.0, 2.0, 3.0, 4.0, 5.0, 6.0)
    arrow = Arrow(line, 0.5)

    filepath = "test_arrow.json"
    arrow.to_json(filepath)

    loaded = Arrow.from_json(filepath)
    assert loaded.radius == 0.5
    assert loaded.mesh.number_of_vertices() == 29
    assert loaded.mesh.number_of_faces() == 28


def test_arrow_mesh_colors():
    """Test that arrow mesh has color collections."""
    line = Line(0.0, 0.0, 0.0, 0.0, 0.0, 10.0)
    arrow = Arrow(line, 1.0)

    assert len(arrow.mesh.pointcolors) == 29
    assert len(arrow.mesh.facecolors) == 28
    assert len(arrow.mesh.linecolors) == 56
    assert len(arrow.mesh.widths) == 56
